dwave/optimization/symbols.pyx

```diff
@@ -2225,6 +2269,77 @@ cdef class LP(Symbol):
+        if self.ptr.variables_shape().size() != 1:
+            raise RuntimeError  # should never happen, but avoid the segfault just in case
+        return self.ptr.variables_shape()[0]
+
+    def _set_state(self, Py_ssize_t index, state):
+        """Set the output of the LP."""
+        if not self.model.is_locked() and self.node_ptr.topological_index() < 0:
+            raise TypeError("the state of an intermediate variable cannot be set without "
+                            "locking the model first. See model.lock().")
+
+        # Convert the state into something we can handle in C++.
+        # This also does some type checking etc
+        cdef double[::1] arr = np.ascontiguousarray(state, dtype=np.double)
+
+        # Reset our state, and check whether that's possible
+        self.reset_state(index)
+
+        # Also make sure our predecessors all have states
+        cdef States states = self.model.states  # for Cython access
+        for pred in self.iter_predecessors():
+            self.model._graph.recursive_initialize(states._states.at(index), (<Symbol>pred).node_ptr)
+
+        # The validity of the state is checked in C++
+        self.ptr.initialize_state(states._states.at(index), _as_span(arr))
+
+    def _states_from_zipfile(self, zf, *, num_states, version):
+        if version < (1, 0):
+            raise ValueError("LP symbol serialization requires serialization version 1.0 or newer")
+
+        # Test whether we have any states saved.
+        try:
+            states_info = zf.getinfo(f"nodes/{self.topological_index()}/states.npy")
+        except KeyError:
+            # No states, so nothing to load
+            return
+
+        # If we have states to load, then go ahead and do so.
+        with zf.open(states_info, mode="r") as f:
+            states = np.load(f, allow_pickle=False)
+
+        for state_index, state in enumerate(states):
+            if np.isnan(state).any():  # we saved missing states with nan
+                continue
+            self._set_state(state_index, state)
+
+    def _states_into_zipfile(self, zf, *, num_states, version):
+        if version < (1, 0):
+            raise ValueError("LP symbol serialization requires serialization version 1.0 or newer")
+
+        # check if there is anything to save, if no then just go ahead and return
+        if not any(self.has_state(i) for i in range(num_states)):
+            return
+
+        # We'll save our states into a dense array. And use NaN to signal when no state
+        # is present.
+        states = np.empty((num_states, self._num_columns()), dtype=np.double)
+        for state_index in range(num_states):
+            if self.has_state(state_index):
+                # we save the state regardless of whether it is feasible or not
+                # In the future we could choose to ignore infeasible states.
+                states[state_index, :] = self.state(state_index)
+            else:
+                raise NotImplementedError
```

Comment:
Why not `states[state_index, :] = np.nan` or similar since deserializing already supports that? Or better just initialize the array as all nan?

Reply:
because I forgot to implement and more importantly to test this branch :laughing: Yes that's what it'll be